File: pix.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <netdb.h>
#include <regex.h>
#include "cJSON.h"
#define ERRMSG "\33[1;31mError: \33[0m"
#define WARNMSG "\33[1;35mWARNING: \33[0m"
#define PORT 80
#define BUFSIZE 2048
#define SOCKBUF 2097152
#define API_H5 "h5.pipix.com"
#define API_PIX "is.snssdk.com"
#define PD(N) printf("length: %zd\n", (N));
#define USER_AGENT "Mozilla/5.0 (Linux; Android 12; XT2201-2) " \
	"AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.57 Mobile Safari/537.36"
#define ACCEPT "Application/json; Charset: UTF-8"
#define CONNECTION "Close"

unsigned long int fn(char *restrict, const char *restrict, const char *restrict);
char *regex(char *restrict, const char *restrict);

int main(int argc, char **argv)
{
	if (argc == 1)
	{
		fprintf(stderr, "%sParameter is little.\n", ERRMSG);
		exit(EXIT_FAILURE);
	}
	char *msg = (char *)malloc(BUFSIZE * sizeof(char));
	char *addr = (char *)malloc(BUFSIZE * sizeof(char));
	char *buffer = (char *)malloc(SOCKBUF * sizeof(char));
	char *buffer_ = (char *)malloc(SOCKBUF * sizeof(char));
	char command[1000] = "am start -a android.intent.action.VIEW -d \"";
	char *ch = NULL;
	cJSON *json = NULL;
	cJSON *obj[4] = {NULL};

	memset(addr, '\0', BUFSIZE * sizeof(char));
	memset(msg, '\0', BUFSIZE * sizeof(char));
	memset(buffer, '\0', SOCKBUF * sizeof(char));
	memset(buffer_, '\0', SOCKBUF * sizeof(char));
	strcpy(addr, regex(argv[1], "/{1}[a-z0-9A-Z]{5,}") + 1);
	sprintf(msg, "GET /s/%s HTTP/1.1\r\n"
		"Host: %s\r\n"
		"User-Agent: %s\r\n"
		"Accept: %s\r\n"
		"Connection: %s\r\n"
		"\r\n"
		"{}",
		addr, API_H5, USER_AGENT, ACCEPT, CONNECTION
	);
	fn(buffer, msg, API_H5);

	memset(msg, '\0', BUFSIZE * sizeof(char));
	memset(addr, '\0', BUFSIZE * sizeof(char));
	strcpy(addr, regex(buffer, "item/?[0-9]{10,}") + 5);
	memset(buffer, '\0', SOCKBUF * sizeof(char));
	sprintf(msg, "GET /bds/cell/detail/?cell_type=1&aid=1319&app_name=super&cell_id=%s HTTP/1.1\r\n"
		"Host: %s\r\n"
		"User-Agent: %s\r\n"
		"Accept: %s\r\n"
		"Connection: %s\r\n"
		"\r\n"
		"{}",
		addr, API_PIX, USER_AGENT, ACCEPT, CONNECTION
	);
	fn(buffer, msg, API_PIX);
	if ((ch = strchr(buffer, '{')))
		strncpy(buffer_, ch, strlen(ch) + 1);
	if ((json = cJSON_Parse(buffer_)) == NULL)
		fprintf(stderr, "%sInitialization json object error\n", ERRMSG);
	if ((*obj = cJSON_GetObjectItem(cJSON_GetObjectItem(json, "data")->child, "item")) == NULL)
		fprintf(stderr, "%sParse json object error\n", ERRMSG);
	if ((obj[1] = cJSON_GetObjectItem(*obj, "content")) == NULL)
		fprintf(stdout, "%sThe video have not title\n", WARNMSG);
	if ((obj[2] = cJSON_GetObjectItem(cJSON_GetArrayItem(cJSON_GetObjectItem(*obj, "comments"), 0), "text")) == NULL)
		fprintf(stderr, "%sThe video have not god comment\n", WARNMSG);
	if ((obj[3] = cJSON_GetObjectItem(
		cJSON_GetArrayItem(
			cJSON_GetObjectItem(
				cJSON_GetObjectItem(
					cJSON_GetObjectItem(
						*obj, "video"
					), "video_high"
				), "url_list"
			), 1
		), "url"
	)) == NULL)
		fprintf(stderr, "%sVideo parse fail.\n", ERRMSG);
	printf("Video Title: %s\n", (obj[1]) ? obj[1]->valuestring : (NULL));
	printf("Video god comment: %s\nVideo url: %s\n", (obj[2]) ? obj[2]->valuestring : (NULL), (obj[3]) ? obj[3]->valuestring : (NULL));
	system(strcat(strcat(command, (obj[3]) ? obj[3]->valuestring : (NULL)), "\""));

	memset(msg, '\0', BUFSIZE * sizeof(char));
	memset(addr, '\0', BUFSIZE * sizeof(char));

	cJSON_Delete(json);
	free(msg);
	free(addr);
	free(buffer);
	free(buffer_);
	return 0;
}

unsigned long int fn(char *restrict buffer, const char *restrict msg, const char *restrict addr)
{
	int sockfd = -1;
	struct sockaddr_in addr_;
	struct hostent *host = NULL;
	long int recvb = 0;
	unsigned long int recvbyte = 0;

	if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0)
		fprintf(stderr, "%sCreate socket-description fail. socket:%d\n", ERRMSG, sockfd);
	// printf("Create socket-file-description suceess: %d\n", sockfd);

	if (!(host = gethostbyname(addr)))
		fprintf(stderr, "%sGet IP of %s fail.\n", ERRMSG, addr);
	memset(&addr_, 0, sizeof(struct sockaddr));
	addr_.sin_family = AF_INET;
	addr_.sin_addr = *((struct in_addr *)host->h_addr_list[0]);
	addr_.sin_port = htons(PORT);

	if ((connect(sockfd, (struct sockaddr *)&addr_, sizeof(struct sockaddr))) < 0)
		fprintf(stderr, "%sConnect to %s of %s fail.\n", ERRMSG, inet_ntoa(addr_.sin_addr), addr);
	// printf("Connect to \33[1;46m%s\33[0m of %s success.\n", inet_ntoa(addr_.sin_addr), addr);
	if ((write(sockfd, msg, strlen(msg))) < 0)
		fprintf(stderr, "%sRequest %s fail.\n", ERRMSG, addr);
	while ((recvb = read(sockfd, buffer + recvbyte, SOCKBUF - recvbyte)) > 0)
		recvbyte += recvb;
	close(sockfd);
	return recvbyte;
}

char *regex(char *restrict string, const char *restrict pattern)
{
	char *buffer = (char *)malloc(SOCKBUF * sizeof(char));
	char ebuf[256] = {'\0'};
	regex_t reg;
	const size_t nmatch = 10;
	regmatch_t pmatch[nmatch];
	int reg_err_code = 0;

	memset(buffer, '\0', SOCKBUF * sizeof(char));
	if ((reg_err_code = regcomp(&reg, pattern, REG_EXTENDED)) != 0)
	{
		regerror(reg_err_code, &reg, ebuf, sizeof(ebuf) / sizeof(char));
		fprintf(stderr, "%s%s: Pattern: %s\n", ERRMSG, ebuf, pattern);
		exit(EXIT_FAILURE);
	}
	if ((reg_err_code = regexec(&reg, string, nmatch, pmatch, 0)) == REG_NOMATCH)
	{
		printf("%s\n%s\n", "Not found result: Your share link is error.", string);
		exit(EXIT_SUCCESS);
	} else if (reg_err_code != 0)
	{
		regerror(reg_err_code, &reg, ebuf, sizeof(ebuf) / sizeof(char));
		fprintf(stderr, "%s%s: Pattern2: %s\n", ERRMSG, ebuf, pattern);
		exit(EXIT_FAILURE);
	}
	for (int i = 0; i < nmatch && pmatch[i].rm_so != -1; i++)
		strncpy(buffer, string + pmatch[i].rm_so, pmatch[i].rm_eo - pmatch[i].rm_so);
	strncpy(string, buffer, strlen(buffer) + 1);
	regfree(&reg);
	free(buffer);
	return string;
}

File: pix.py
# -*- coding: UTF-8 -*-
from flask import Flask, request
import requests
import json
import re

class parse(object):
	def __init__(self):
		self.user_agent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.5060.114 Safari/537.36 Edg/103.0.1264.62"
		self.accept = "application/json; charet: UTF-8"
		self.connection = "Close"
		self.content_type = "application/x-www-form-urlencoded"

	def pipix(self, id):
		id_ = re.compile("(/{1}s/{1})([0-9a-zA-Z]{5,})").findall(id)
		response = requests.get(
			url = "https://h5.pipix.com/s/" + id_[0][1],
			headers = {
				"Host": "h5.pipix.com",
				"User-Agent": self.user_agent,
				"Accept": self.accept,
				"Connection": self.connection
			},
			allow_redirects = False
		)
		id_ = re.compile("(/{1}item/{1})([0-9]{10,})").findall(response.text)
		response = json.loads(requests.get(
			url = "https://is.snssdk.com/bds/cell/detail/?cell_type=1&aid=1319&app_name=super&cell_id=" + id_[0][1],
			headers = {
				"Host": "is.snssdk.com",
				"User-Agent": self.user_agent,
				"Accept": self.accept,
				"Connection": self.connection
			},
			allow_redirects = False
		).text)
		god_comment = None
		if response["data"]["data"]["item"]["comments"]:
			god_comment = response["data"]["data"]["item"]["comments"][0]["text"]
		return {
			"message": "success",
			"data": {
				"title": response["data"]["data"]["item"]["content"],
				"god-comment": god_comment,
				"url": response["data"]["data"]["item"]["video"]["video_high"]["url_list"][0]["url"]
			}
		}

	def kuaishou(self, id):
		id_ = re.compile("((http|https)?://v.(kuaishou|kuaishouapp)?\.com/.{3,6}[a-z0-9A-Z]+)").findall(id)
		response = requests.get(
			url = id_[0][0],
			headers = {
				"Host": "v.kuaishou.com",
				"User-Agent": self.user_agent,
				"Accept": self.accept,
				"Connection": self.connection
			},
			allow_redirects = False
		)
		id_ = re.compile("photoId\=([0-9a-zA-Z]{10,})").findall(response.headers["Location"])
		response = json.loads(requests.post(
			url = "https://v.m.chenzhongtech.com/rest/wd/photo/info",
			data = json.dumps({
				"photoId": str(id_[0]),
				"isLongVideo": "false"
			}),
			headers = {
				"Host": "v.m.chenzhongtech.com",
				"User-Agent": self.user_agent,
				"Accept": self.accept,
				"Connection": "Keep-Alive",
				"Content-Type": "application/json",
				"Origin": "https://www.kuaishou.com",
				"Referer": response.headers["Location"],
				"Cookie": "did=web_5f1471502c301bfca2385b8de925aa69; didv=1658645570591;"
			},
			allow_redirects = False
		).text)
		return {
			"message": "success",
			"data": {
				"title": response["shareInfo"]["shareTitle"],
				"god-comment": None,
				"url": response["photo"]["mainMvUrls"][0]["url"]
			}
		}

	def douyin(self, id):
		id_ = re.findall(re.compile("((http|https)?://v\.douyin\.com/[a-z0-9A-Z]+)"), id)
		response = requests.get(
			url = id_[0][0],
			headers = {
				"Host": "v.douyin.com",
				"User-Agent": self.user_agent,
				"Accept": self.accept,
				"Connection": self.connection
			},
			allow_redirects = False
		)
		id_ = re.findall(re.compile("video/([0-9]+)/.+"), response.headers["Location"])
		response = json.loads(requests.get(
			url = "https://www.iesdouyin.com/web/api/v2/aweme/iteminfo/?item_ids=" + id_[0],
			headers = {
				"Host": "www.iesdouyin.com",
				"User-Agent": self.user_agent,
				"Accept": self.accept,
				"Connection": self.connection
			},
			allow_redirects = False
		).text)
		return {
			"message": "success",
			"data": {
				"title": response["item_list"][0]["desc"],
				"god-comment": None,
				"url": response["item_list"][0]["video"]["play_addr"]["url_list"][0]
			}
		}

	def run(self):
		# a = self.pipix("https://h5.pipix.com/s/251aans/# [皮皮虾] 孤独就像这个圈圈 【长按复制】到浏览器观看")
		# b = self.kuaishou("https://v.kuaishouapp.com/s/QKtYIfi3 卢卢溜溜球 \"可莉 \"仲夏幻夜奇想曲 \"原神 复制此消息，打开【快手极速版】直接观看！")
		# b = self.kuaishou("https://v.kuaishou.com/qZgvoL 你以为元歌那么好玩吗？我也来试试针对元歌")
		c = self.douyin("7.48 qRk:/ 复制打开抖音，看看【小纯游戏（原神）的作品】# 仲夏幻夜奇想曲 # 原神 # 原神与你同行 可怜的霄... https://v.douyin.com/2cHLCwy/")
		print("Video title:", c["data"]["title"])
		print("Video god comment:", c["data"]["god-comment"])
		print("Video url:", c["data"]["url"])
		# print("====\t====")
		# print("Video title:", b["data"]["title"])
		# print("Video url:", b["data"]["url"])

app = Flask(__name__)
sdk250 = parse()

@app.route("/", methods = ["GET", "POST"])
def index():
	url = request.args.get("url")
	if url:
		if re.findall(re.compile("(kuaishou|kuaishouapp)"), url) != []:
			return sdk250.kuaishou(url)
		elif re.findall(re.compile("(pipix)"), url) != []:
			return sdk250.pipix(url)
		elif re.findall(re.compile("(douyin)"), url) != []:
			return sdk250.douyin(url)
		else:
			return {
				"code": 404,
				"message": "Null"
			}
	else:
		return "URL is Null.\n"

if __name__ == "__main__":
	app.run(host = "0.0.0.0", port = 20880, debug = False)
